packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java

```diff
@@ -850,6 +937,158 @@ public void downloadResumablePauseAsync(final String uuid, final Promise promise
     }
   }
 
+  @ExpoMethod
+  public void readSAFDirectoryAsync(String uriStr, Map<String, Object> options, Promise promise) {
+    try {
+      Uri uri = Uri.parse(uriStr);
+      ensurePermission(uri, Permission.READ);
+      if (isSAFUri(uri)) {
+        DocumentFile file = DocumentFile.fromTreeUri(getContext(), uri);
+        if (file == null || !file.exists() || !file.isDirectory()) {
+          promise.reject("ERR_FILESYSTEM_CANNOT_READ_DIRECTORY",
+            "Uri '" + uri + "' doesn't exist or isn't a directory.");
+          return;
+        }
+        DocumentFile[] children = file.listFiles();
+        List<String> result = new ArrayList<>();
+        for (DocumentFile child : children) {
+          result.add(child.getUri().toString());
+        }
+        promise.resolve(result);
+      } else {
+        throw new IOException("Unsupported scheme for location '" + uri + "'.");
```

Comment:
```suggestion
        throw new IOException("The URI '" + uri + "' is not a Storage Access Framework URI. Try using FileSystem.readDirectoryAsync instead.");
```
etc. for the rest of these functions